Add API endpoint to request contents of multiple files simultaniously

denyskon opened the pull request:
Adds an API POST endpoint under `/repos/{owner}/{repo}/file-contents` which receives a list of paths and returns a list of the contents of these files.

This API endpoint will be helpful for applications like headless CMS (reference: https://github.com/sveltia/sveltia-cms/issues/198) which need to retrieve a large number of files by reducing the amount of needed API calls.

Close #33495

wxiaoguang commented on routers/api/v1/api.go:
```diff
@@ -1382,6 +1382,7 @@ func Routes() *web.Router {
 				m.Group("/contents", func() {
 					m.Get("", repo.GetContentsList)
 					m.Post("", reqToken(), bind(api.ChangeFilesOptions{}), reqRepoBranchWriter, mustNotBeArchived, repo.ChangeFiles)
+					m.Post("/files", bind(api.GetFilesOptions{}), repo.GetContentsFiles)
```
I do not think it could use this endpoint. It conflicts with the `Post("/*")` below.

denyskon replied:
Removed `/contents/` prefix, now it's just `/repos/owner/repo/files`

wxiaoguang commented:
Some thoughts in my mind:

1. Why it is a "POST" endpoint? If I understand correctly, it "GETs"
2. Does GitHub have a simliar API? If yes, we'd better to follow, if no, let's invent
3. It needs some tests (almost all existing APIs are covered by tests, otherwise it would easily break in the future if something changes)

denyskon commented:
1. IIRC, the HTTP spec says or has said that GET requests should not contain a body - that's why it isn't an option. Post is the next best choice I think.

2. GitHub doesn't have a similar API, as such functionality is now being served through their GraphQL API. This, however, is clearly out of scope for near future, so I think that such an API endpoint would be helpful :)

3. I can add some tests...

silverwind commented:
> 1. IIRC, the HTTP spec says or has said that GET requests should not contain a body - that's why it isn't an option. Post is the next best choice I think.

[HTTP QUERY](https://httpwg.org/http-extensions/draft-ietf-httpbis-safe-method-w-body.html) is a proposed standard to solve the GET-with-body problem. Tooling support is probably not there yet, but if it works in your case, it might be an option to use it already.

wxiaoguang reviewed (commented):
> 1. IIRC, the HTTP spec says or has said that GET requests should not contain a body - that's why it isn't an option. Post is the next best choice I think.

I see, I missed the `Files []string` field, it would be quite large so it is POST's job.

denyskon commented:
@wxiaoguang Added an integration test

lunny reviewed (approved):
Except https://github.com/go-gitea/gitea/pull/34139/files#r2040159364

denyskon commented on services/repository/files/file.go:
```diff
@@ -17,12 +17,17 @@ import (
+	return files
+}
+
+func GetFilesResponseFromCommit(ctx context.Context, repo *repo_model.Repository, commit *git.Commit, branch string, treeNames []string) (*api.FilesResponse, error) {
```
Removed

wxiaoguang commented:
Requesting 100 files with 10MB each will consume more than 1.3G memory and lead to DoS attack to the server?

denyskon commented:
@wxiaoguang Isn't that concern resolved by pagination?

wxiaoguang commented:
> @wxiaoguang Isn't that concern resolved by pagination?

1. I do not see the `pagination` is right. You just loaded everything into memory and then `PaginateSlice`. OOM occurs before `pagination`
2. What if 10 files, each 100MB?

denyskon commented:
@wxiaoguang 

About 1. you were right, I now moved pagination to be applied before retrieving the files

About 2: How is it different to requesting one 1GB large file over the existing file API? If you put it like that, every kind of API endpoint is dangerous to publish...

wxiaoguang commented:
> About 2: How is it different to requesting one 1GB large file over the existing file API? If you put it like that, every kind of API endpoint is dangerous to publish...

I do not think it is true. At least, there are some limits in code.

denyskon commented:
Do you have any suggestions how it could be made safer? I honestly do not see a way - if there's already something in the code for other endpoints, please point me to it, but otherwise I don't see a reason for exactly this PR to be treated differently.

wxiaoguang commented:
Actually I just hope you could look into code .... the answer is in the code. For example: `GetContents` -> `GetBlobBySHA` -> `gitBlob.Size() <= setting.API.DefaultMaxBlobSize`.


And in this case, I do not think some errors could be "ignored" as old code does. If the API caller gets empty file content response (internal error, or size exceeded), would it go wrong? If it is an important API, some edge cases (especially the too-large files) should be covered by tests.

denyskon commented:
@wxiaoguang Thanks for pointing out, I in fact didn't notice this mechanism. I implemented a similar, configurable limit for the combined size of the response.

denyskon commented:
> I also removed the `// ok if fails, then will be nil` and replaced it by a 404 error

@wxiaoguang As the method this API endpoint is based on is being used by other API endpoints this change would be breaking (see https://github.com/go-gitea/gitea/actions/runs/14539074128), so I reverted this change. I don't think we should break existing API workflows

wxiaoguang commented:
Thank you for the updates. I have some questions about the design:

1. What if something wrong happens in edge cases, what client would see and what they could do?

Some examples, assume "client requests File1...File100, server PageSize=25", then in these cases, what client would see and could do?

    1. At File3, File3 exceeds the MaxBlobSize
    2. At File4, File4 doesn't exist
    3. At File5, File5 causes git error (exists but just server internal error)
    4. At File8, the response exceeds the MaxResponseSize

2. Is `util.PaginateSlice(files, listOpts.Page, listOpts.PageSize)` the proper way to handle client requests?
    * By this design: client needs to keep posting File1....File100 with different "page" number.
    * Maybe other approaches could be like: client requests File1...File100, server responds File1...File17 as much as possible, then client request File18...File100, then server responds File18...File99 as much as possible.
    * Which approach is more friendly to API clients?

denyskon commented:
The response would be as follows:

1. This is being handled by the mechanism of `GetBlobBySHA` - the returned file just hast `content: ""` then. I don't think it is right, but a change of this behaviour would be breaking to other API endpoints.

2. & 3. If File4 does not exist, file 4 will be `null` instead of the returned file inside the list. The same is with a Git error on File5, it will also be `null`. As stated above, a change to this behaviour would also be breaking to other API endpoints.

4. If the response exceeds the max response size, an error 413 with an error message stating that mac response size was exceeded will be returned.

denyskon commented:
Talking about pagination, I would keep the page number approach as it is the same as in other API endpoints which use pagination. Of course the difference here is that there's a request body which sets the required logic, but still I would keep the similarity to the rest of our API as much as possible.

I also think that it is easier for the client to just keep sending the whole list and increment one number than to change the file list each time, isn't it?

wxiaoguang commented:
> The response would be as follows:

Thank you, I can see "what would happen" now.

But, what the client "could" do? For example: 

* `At File3, File3 exceeds the MaxBlobSize` -> `the returned file just hast content: ""`, then client thinks that it is an empty file, using an "empty" file would cause wrong result? Even more, if there is any modification to it and upload to the server, the the file will be broken?
* `At File8, the response exceeds the MaxResponseSize` -> `an error 413 with an error message`, then you have current "pagination"  behavior, then how would the client read the remaining files? How could client read all files correctly?
* The existing design for "exceeding limit case" is: provide a blob link and let client read the content by download stream.

> I also think that it is easier for the client to just keep sending the whole list and increment one number than to change the file list each time, isn't it?

For a single point, any design is fine. But when we combine the factors together, is it still the good design? See the "File8" case above: how could a client get all files correctly if one page exceeds the limit?

denyskon commented:
If we stick with this pagination method, you would just send the request starting with the first non-responded file using a smaller page size - but I see that it isn't very intuitive.

However, if we go with your suggestion, we need to find out how many files are returned per request. As many as per the default page size, but less if the page limit is reached? Then we would also set the total count header, and the way to go for the client would be: if the number in the header is larger than the returned number, re-request the last (count-returned_count) files. That shouldn't be problematic, but we need to document it properly.

If the blob is too large, there's still the normal metadata returned so that you can use the direct link. We also return file size, so if `size > 0` but `content: ""`, the user also knows that he has to re-request - but we also need to document this.

wxiaoguang commented:
> Then we would also set the total count header,

I do not think this "total count header" is useful to caller. The caller already knows the "total count" because the file list was just sent by itself.

----


> but we also need to document this.

Not only document, I think it needs tests to make the API implementation stable and won't break.

wxiaoguang commented:
Actually there could be another approach: when the response exceeds the "response limit", omit all the following files' contents as they also exceed the blob limit. For example (the sample above):

* File1: has content
* File2: has content
* File3: no content (because it exceeds max blob size)
* File4: no content (because it does not exist, and maybe some other fields could help to distinguish)
* File5: no content (because it causes error, and maybe some other fields could help to distinguish)
* File6: has content
* File7: has content
* File8: no content (because it makes the response exceed the max response size)
* File9: has content (suppose it is small enough)
....

denyskon commented:
@wxiaoguang I now went with the first of the suggested methods - if the response would exceed the max size, the list is being aborted at that point and the user needs to request again for the next files. It is documented inside swagger and I also added a test for pagination with small files, with large files (doesn't reach default paging num because of response size limit) and for too large file (content empty).

wxiaoguang commented:
Found various legacy (related) problems, need to fix together. If I open a separate refactoring PR, the PRs would conflict largely. So I made the changes in this PR directly. 🙏

denyskon commented:
@wxiaoguang Looks good, I already noticed that the mechanism for getting the ref was flawed. I also missed that base64 encoding increases the size - I adapted the test to it. Could you fix the lint issues in your bugfix?

wxiaoguang commented:
One more thing, it's impossible to support make "POST" work with "read" permission at the moment, so I also added the "GET" method support. It is somewhat hacky but I think it is the only thing we can do at the moment.
